### PR TITLE
Refactor endpoints to be Smithy-native

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
- #1637 

## Description
This makes a 99% backwards compatible refactoring to:
1. Create the `aws_smithy_types::Endpoint` type
2. Create the `aws_smithy_http::endpoint::ResolveEndpoint` trait (tbd: this may move into aws_smithy_types)
3. Use `ResolveEndpoint` inside of `<service>::Config` internally instead of `ResolveAwsEndpoint`
4. Shim `ResolveAwsEndpoint` against `ResolveEndpoint` via a static `Params` struct (in the future, this will be code generated per-service).
5. Update the `make_operation` to write the endpoint into the bag instead of writing the endpoint resolver into the bag & update the AWS middleware accordingly

## Testing
- [x] UT/IT
- [ ] canary

## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws/rust-runtime/aws-endpoint/src/lib.rs [hunk 1]
```diff
@@ -10,27 +10,79 @@ pub mod partition;
...
+}
+impl ResolveEndpoint<Params> for EndpointShim {
+    fn resolve_endpoint(
+        &self,
+        params: &Params,
+    ) -> Result<SmithyEndpoint, aws_smithy_http::endpoint::Error> {
+        let aws_endpoint = self
+            .0
+            .resolve_endpoint(params.region.as_ref().unwrap_or(
+                /*EndpointError::message("no region in params")*/
```
Comment (Velfi) [inline aws/rust-runtime/aws-endpoint/src/lib.rs:63]: Leftover comment?

File: aws/rust-runtime/aws-endpoint/src/lib.rs [hunk 2]
```diff
@@ -56,37 +108,95 @@ impl Display for AwsEndpointStageError {
...
+             */
+            let endpoint_result = props
+                .remove::<aws_smithy_http::endpoint::Result>()
+                .ok_or(AwsEndpointStageError::NoEndpointResolver)?;
+            let endpoint = endpoint_result.map_err(|err|AwsEndpointStageError::EndpointResolutionError(err.into()))?;
+            props.insert::<aws_smithy_http::endpoint::Result>(Ok(endpoint));
+            // unwrap safety: we just put it in the bag, we can take it out again.
+            let endpoint = props
+                .get::<aws_smithy_http::endpoint::Result>()
+                .map(|res|res.as_ref().unwrap()).unwrap();
```
Comment (Velfi) [inline aws/rust-runtime/aws-endpoint/src/lib.rs:133]: Is leaving a comment like this preferable to using `.expect()`?

**Review by Velfi — APPROVED** [review]

File: aws/rust-runtime/aws-endpoint/src/lib.rs (hunk 1 above)
Comment (rcoh) [reply aws/rust-runtime/aws-endpoint/src/lib.rs]: good catch, fixed

File: rust-runtime/aws-smithy-http/src/endpoint.rs [hunk 3]
```diff
@@ -48,6 +92,28 @@ pub enum InvalidEndpoint {
     EndpointMustHaveAuthority,
 }
 
+pub fn apply_endpoint(uri: &mut Uri, endpoint: &Uri, prefix: Option<&EndpointPrefix>) {
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-http/src/endpoint.rs:95]: Needs documentation

File: rust-runtime/aws-smithy-types/src/endpoint.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    }
+
+    /// Constructs an [`Endpoint`] from this builder
+    ///
+    /// # Panics
+    /// Panics if URL is unset or empty
+    pub fn build(self) -> Endpoint {
+        if self.endpoint.url() == "" {
+            panic!("URL was unset");
+        }
```
Comment (jdisanti) [inline rust-runtime/aws-smithy-types/src/endpoint.rs:125]:
```suggestion
        assert!(self.endpoint.url() != "", "URL was unset");
```

**Review by jdisanti — APPROVED** [review]
LGTM!

**Pull request merged**